api/v1/pxctemplate_types.go

```diff
@@ -0,0 +1,45 @@
+// dbaas-operator
+// Copyright (C) 2022 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	pxcv1 "github.com/percona/percona-xtradb-cluster-operator/pkg/apis/pxc/v1"
+)
+
+//+kubebuilder:object:root=true
+
+// PXCTemplate is the Schema for the pxctemplates API
+type PXCTemplate struct {
```

Comment:
https://redhat-developer.github.io/service-binding-operator/userguide/binding-workloads-using-sbo/binding-options.html#composing-custom-binding-data

https://redhat-developer.github.io/service-binding-operator/userguide/binding-workloads-using-sbo/custom-path-injection.html#workload-resource-mapping

or we can have some middle ground if we want to have a new object type - some general DBTemplate type and a specific kind inside and name.

Reply:
I'm not sure we can achieve the intended behaviour without having a new type.
(Using PXC as an example next but it could be any other) If we simply have `dbaas.percona.com/dbtemplate-kind` and `dbaas.percona.com/dbtemplate-name` referencing a `PerconaXtraDBCluster` CR to be used as a template it means that that `PerconaXtraDBCluster` object must exist in k8s key-value store already and this would lead to the PXC operator creating that DBInstance. i.e. you could only use already existing DB instances as templates and you'd not be able to simply create a template without creating a new DB instance.
Am I missing something in the design that avoids this unwanted behaviour?

Reply:
So I just went through the proposal document again and I think I now understand what you mean with
> The idea of the original proposal was to don't introduce new types

You're not saying that the types don't need to exist (what I was arguing against in my last comment), you're saying that it's not the dbaas-operator's responsibility to create them. It's up to end-user to ensure that these template CRDs exist, like in the following example.
https://github.com/percona/dbaas-operator/blob/861edb88d886c50381c061b2ac897c0982802897/docs/proposals/001-DatabaseCluster-Templates/proposal.md?plain=1#L172-L195

Am I understanding this correctly?

Reply:
> I'm not sure we can achieve the intended behaviour without having a new type. (Using PXC as an example next but it could be any other) If we simply have `dbaas.percona.com/dbtemplate-kind` and `dbaas.percona.com/dbtemplate-name` referencing a `PerconaXtraDBCluster` CR to be used as a template it means that that `PerconaXtraDBCluster` object must exist in k8s key-value store already and this would lead to the PXC operator creating that DBInstance. i.e. you could only use already existing DB instances as templates and you'd not be able to simply create a template without creating a new DB instance. Am I missing something in the design that avoids this unwanted behaviour?

So I think we should clarify this: 
>If one of those two parameters (kind, name) is not set - dbaas-operator wouldn't be able to identify the template and thus would ignore it.

not only if the field doesn't-exist or is empty, but also if Kind/API is not there ( kubectl get ($dbaas.percona.com/dbtemplate-kind) ) or there is no object with such a name - we just should skip.

> `dbaas.percona.com/dbtemplate-name` referencing a `PerconaXtraDBCluster` CR 
it should be some Template Kind, not PXC :) and if that Template Kind or whatever there doesn't exist - skip it :)

> you'd not be able to simply create a template
hm, that's where the confusion comes from. CRDs for templates are separate from the DBInstances and they are applied manually and don't have any controller behind them.
https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/
kind of CronJob here. It is just another object in k8s without hooks behind it.

So Templates could exist in a cluster even without any operator installed or any DBInstance. It is up to SRE/DBA when and how to apply CRD for them and create CRs

Reply:
> So I just went through the proposal document again and I think I now understand what you mean with
> 
> > The idea of the original proposal was to don't introduce new types
> 
> You're not saying that the types don't need to exist (what I was arguing against in my last comment), you're saying that it's not the dbaas-operator's responsibility to create them. It's up to end-user to ensure that these template CRDs exist, like in the following example.
> 
> https://github.com/percona/dbaas-operator/blob/861edb88d886c50381c061b2ac897c0982802897/docs/proposals/001-DatabaseCluster-Templates/proposal.md?plain=1#L172-L195
> 
> Am I understanding this correctly?

correct! so we even shouldn't bother - check if the object of this kind and this name exist - and try to merge it. If object or api/kind doesn't exist you will receive and error and just log it and skip :)

It is up to SRE/DBA manually create it, or it could be some app in a future that creates and manage them.